# This file is part of HiGFXback

# requires
REQUIRES="premake-build stb-build"

pkg-config --exists --print-errors $REQUIRES || exit 1

# configure
rm -f example/stb_image_write.h
premake5 --stb-dir=/include/stb --without-glfw gmake

# build
make -C build

# install
install -d $DESTDIR/bin
install build/example2 $DESTDIR/bin/nanosvg-dec
install -d $DESTDIR/include/nanosvg/src
install -m 644 src/nanosvg.h $DESTDIR/include/nanosvg/src
ln -sf src/nanosvg.h $DESTDIR/include/nanosvg/nanosvg.h
install -m 644 src/nanosvgrast.h $DESTDIR/include/nanosvg/src
ln -sf src/nanosvgrast.h $DESTDIR/include/nanosvg/nanosvgrast.h

# build.pc
install -d $DESTDIR/share/pkgconfig
cat > $DESTDIR/share/pkgconfig/nanosvg-build.pc << EOF
Name: nanosvg
Version: 20241219
Description: single-file SVG parser
Requires: $REQUIRES

devel=\\
/include/nanosvg/nanosvg.h \\
/include/nanosvg/nanosvgrast.h \\
/include/nanosvg/src/nanosvg.h \\
/include/nanosvg/src/nanosvgrast.h

exec=\\
/bin/nanosvg-dec
EOF

exit

# patch
--- nanosvg-20241219.orig/example/example2.c
+++ nanosvg-20241219/example/example2.c
@@ -26,16 +26,20 @@
 #define NANOSVGRAST_IMPLEMENTATION
 #include "nanosvgrast.h"
 
-int main()
+int main(int argc, char *argv[])
 {
 	NSVGimage *image = NULL;
 	NSVGrasterizer *rast = NULL;
 	unsigned char* img = NULL;
 	int w, h;
-	const char* filename = "../example/23.svg";
 
-	printf("parsing %s\n", filename);
-	image = nsvgParseFromFile(filename, "px", 96.0f);
+	if (argc != 3) {
+		fprintf(stderr, "Usage: nanosvg-dec <input-image> <output-image>\n");
+		return 1;
+	}
+
+	printf("parsing %s\n", argv[1]);
+	image = nsvgParseFromFile(argv[1], "px", 96.0f);
 	if (image == NULL) {
 		printf("Could not open SVG image.\n");
 		goto error;
@@ -58,8 +62,8 @@
 	printf("rasterizing image %d x %d\n", w, h);
 	nsvgRasterize(rast, image, 0,0,1, img, w, h, w*4);
 
-	printf("writing svg.png\n");
- 	stbi_write_png("svg.png", w, h, 4, img, w*4);
+	printf("writing %s\n", argv[2]);
+ 	stbi_write_png(argv[2], w, h, 4, img, w*4);
 
 error:
 	nsvgDeleteRasterizer(rast);
--- nanosvg-20241219.orig/premake4.lua
+++ nanosvg-20241219/premake4.lua
@@ -3,54 +3,59 @@
 
 solution "nanosvg"
 	location ( "build" )
-	configurations { "Debug", "Release" }
+	configurations { "Release", "Debug" }
 	platforms {"native", "x64", "x32"}
+
+	newoption { trigger = "prefix", value = "PATH", description = "Installation prefix", default = "/usr/local" }
+	newoption { trigger = "stb-dir", value = "PATH", description = "Directory for stb image headers" }
+	newoption { trigger = "without-glfw", description = "Do not use GLFW" }
+
+	includedirs { "src", _OPTIONS["stb-dir"] }
+	targetdir("build")
   
+	if not _OPTIONS["without-glfw"] then
+
 	project "example1"
 		kind "ConsoleApp"
-		language "C++"
+		language "C"
 		files { "example/example1.c", "example/*.h", "src/*.h" }
-		includedirs { "example", "src" }
-		targetdir("build")
+		buildoptions { "-I" .. _OPTIONS["prefix"] .. "/include"}
+		linkoptions { "-L" .. _OPTIONS["prefix"] .. "/lib", "-Wl,-rpath," .. _OPTIONS["prefix"] .. "/lib" }
 	 
-		configuration { "linux" }
-			 links { "X11","Xrandr", "rt", "GL", "GLU", "pthread", "glfw" }
+		filter { "system:linux" }
+			 links { "GL", "glfw", "m" }
 
-		configuration { "windows" }
-			 links { "glu32","opengl32", "gdi32", "winmm", "user32" }
+		filter { "system:windows" }
+			 links { "opengl32", "glfw3" }
 
-		configuration { "macosx" }
+		filter { "system:macosx" }
 			links { "glfw3" }
 			linkoptions { "-framework OpenGL", "-framework Cocoa", "-framework IOKit", "-framework CoreVideo" }
 
-		configuration "Debug"
+		filter "Debug"
 			defines { "DEBUG" }
-			flags { "Symbols", "ExtraWarnings"}
+			symbols "On"
+			warnings "Extra"
 
-		configuration "Release"
+		filter "Release"
 			defines { "NDEBUG" }
-			flags { "Optimize", "ExtraWarnings"}    
+			optimize "On"
+			warnings "Extra"
+
+	end
 
 	project "example2"
 		kind "ConsoleApp"
-		language "C++"
+		language "C"
 		files { "example/example2.c", "example/*.h", "src/*.h" }
-		includedirs { "example", "src" }
-		targetdir("build")
+		links { "m" }
 	 
-		configuration { "linux" }
-			 links { "X11","Xrandr", "rt", "pthread" }
-
-		configuration { "windows" }
-			 links { "winmm", "user32" }
-
-		configuration { "macosx" }
-			linkoptions { "-framework Cocoa", "-framework IOKit" }
-
-		configuration "Debug"
+		filter "Debug"
 			defines { "DEBUG" }
-			flags { "Symbols", "ExtraWarnings"}
+			symbols "On"
+			warnings "Extra"
 
-		configuration "Release"
+		filter "Release"
 			defines { "NDEBUG" }
-			flags { "Optimize", "ExtraWarnings"}    
+			optimize "On"
+			warnings "Extra"
--- nanosvg-20241219.orig/src/nanosvgrast.h
+++ nanosvg-20241219/src/nanosvgrast.h
@@ -46,6 +46,9 @@
 	unsigned char* img = malloc(w*h*4);
 	// Rasterize
 	nsvgRasterize(rast, image, 0,0,1, img, w, h, w*4);
+
+	// For non-square X,Y scaling, use
+	nsvgRasterizeXY(rast, image, 0,0,1,1, img, w, h, w*4);
 */
 
 // Allocated rasterizer context.
@@ -64,6 +67,12 @@
 				   NSVGimage* image, float tx, float ty, float scale,
 				   unsigned char* dst, int w, int h, int stride);
 
+// As above, but allow X and Y axes to scale independently for non-square aspects
+void nsvgRasterizeXY(NSVGrasterizer* r,
+				     NSVGimage* image, float tx, float ty,
+				     float sx, float sy,
+				     unsigned char* dst, int w, int h, int stride);
+
 // Deletes rasterizer context.
 void nsvgDeleteRasterizer(NSVGrasterizer*);
 
@@ -371,7 +380,7 @@
 	nsvg__flattenCubicBez(r, x1234,y1234, x234,y234, x34,y34, x4,y4, level+1, type);
 }
 
-static void nsvg__flattenShape(NSVGrasterizer* r, NSVGshape* shape, float scale)
+static void nsvg__flattenShape(NSVGrasterizer* r, NSVGshape* shape, float sx, float sy)
 {
 	int i, j;
 	NSVGpath* path;
@@ -379,13 +388,13 @@
 	for (path = shape->paths; path != NULL; path = path->next) {
 		r->npoints = 0;
 		// Flatten path
-		nsvg__addPathPoint(r, path->pts[0]*scale, path->pts[1]*scale, 0);
+		nsvg__addPathPoint(r, path->pts[0]*sx, path->pts[1]*sy, 0);
 		for (i = 0; i < path->npts-1; i += 3) {
 			float* p = &path->pts[i*2];
-			nsvg__flattenCubicBez(r, p[0]*scale,p[1]*scale, p[2]*scale,p[3]*scale, p[4]*scale,p[5]*scale, p[6]*scale,p[7]*scale, 0, 0);
+			nsvg__flattenCubicBez(r, p[0]*sx,p[1]*sy, p[2]*sx,p[3]*sy, p[4]*sx,p[5]*sy, p[6]*sx,p[7]*sy, 0, 0);
 		}
 		// Close path
-		nsvg__addPathPoint(r, path->pts[0]*scale, path->pts[1]*scale, 0);
+		nsvg__addPathPoint(r, path->pts[0]*sx, path->pts[1]*sy, 0);
 		// Build edges
 		for (i = 0, j = r->npoints-1; i < r->npoints; j = i++)
 			nsvg__addEdge(r, r->points[j].x, r->points[j].y, r->points[i].x, r->points[i].y);
@@ -735,7 +744,7 @@
 	}
 }
 
-static void nsvg__flattenShapeStroke(NSVGrasterizer* r, NSVGshape* shape, float scale)
+static void nsvg__flattenShapeStroke(NSVGrasterizer* r, NSVGshape* shape, float sx, float sy)
 {
 	int i, j, closed;
 	NSVGpath* path;
@@ -743,15 +752,16 @@
 	float miterLimit = shape->miterLimit;
 	int lineJoin = shape->strokeLineJoin;
 	int lineCap = shape->strokeLineCap;
-	float lineWidth = shape->strokeWidth * scale;
+	float sw = (sx + sy) / 2;
+	float lineWidth = shape->strokeWidth * sw;
 
 	for (path = shape->paths; path != NULL; path = path->next) {
 		// Flatten path
 		r->npoints = 0;
-		nsvg__addPathPoint(r, path->pts[0]*scale, path->pts[1]*scale, NSVG_PT_CORNER);
+		nsvg__addPathPoint(r, path->pts[0]*sx, path->pts[1]*sy, NSVG_PT_CORNER);
 		for (i = 0; i < path->npts-1; i += 3) {
 			float* p = &path->pts[i*2];
-			nsvg__flattenCubicBez(r, p[0]*scale,p[1]*scale, p[2]*scale,p[3]*scale, p[4]*scale,p[5]*scale, p[6]*scale,p[7]*scale, 0, NSVG_PT_CORNER);
+			nsvg__flattenCubicBez(r, p[0]*sx,p[1]*sy, p[2]*sx,p[3]*sy, p[4]*sx,p[5]*sy, p[6]*sx,p[7]*sy, 0, NSVG_PT_CORNER);
 		}
 		if (r->npoints < 2)
 			continue;
@@ -797,7 +807,7 @@
 				dashOffset -= shape->strokeDashArray[idash];
 				idash = (idash + 1) % shape->strokeDashCount;
 			}
-			dashLen = (shape->strokeDashArray[idash] - dashOffset) * scale;
+			dashLen = (shape->strokeDashArray[idash] - dashOffset) * sw;
 
 			for (j = 1; j < r->npoints2; ) {
 				float dx = r->points2[j].x - cur.x;
@@ -819,7 +829,7 @@
 					// Advance dash pattern
 					dashState = !dashState;
 					idash = (idash+1) % shape->strokeDashCount;
-					dashLen = shape->strokeDashArray[idash] * scale;
+					dashLen = shape->strokeDashArray[idash] * sw;
 					// Restart
 					cur.x = x;
 					cur.y = y;
@@ -988,7 +998,7 @@
 }
 
 static void nsvg__scanlineSolid(unsigned char* dst, int count, unsigned char* cover, int x, int y,
-								float tx, float ty, float scale, NSVGcachedPaint* cache)
+								float tx, float ty, float sx, float sy, NSVGcachedPaint* cache)
 {
 
 	if (cache->type == NSVG_PAINT_COLOR) {
@@ -1029,9 +1039,9 @@
 		int i, cr, cg, cb, ca;
 		unsigned int c;
 
-		fx = ((float)x - tx) / scale;
-		fy = ((float)y - ty) / scale;
-		dx = 1.0f / scale;
+		fx = ((float)x - tx) / sx;
+		fy = ((float)y - ty) / sy;
+		dx = 1.0f / sx;
 
 		for (i = 0; i < count; i++) {
 			int r,g,b,a,ia;
@@ -1074,9 +1084,9 @@
 		int i, cr, cg, cb, ca;
 		unsigned int c;
 
-		fx = ((float)x - tx) / scale;
-		fy = ((float)y - ty) / scale;
-		dx = 1.0f / scale;
+		fx = ((float)x - tx) / sx;
+		fy = ((float)y - ty) / sy;
+		dx = 1.0f / sx;
 
 		for (i = 0; i < count; i++) {
 			int r,g,b,a,ia;
@@ -1115,7 +1125,7 @@
 	}
 }
 
-static void nsvg__rasterizeSortedEdges(NSVGrasterizer *r, float tx, float ty, float scale, NSVGcachedPaint* cache, char fillRule)
+static void nsvg__rasterizeSortedEdges(NSVGrasterizer *r, float tx, float ty, float sx, float sy, NSVGcachedPaint* cache, char fillRule)
 {
 	NSVGactiveEdge *active = NULL;
 	int y, s;
@@ -1197,7 +1207,7 @@
 		if (xmin < 0) xmin = 0;
 		if (xmax > r->width-1) xmax = r->width-1;
 		if (xmin <= xmax) {
-			nsvg__scanlineSolid(&r->bitmap[y * r->stride] + xmin*4, xmax-xmin+1, &r->scanline[xmin], xmin, y, tx,ty, scale, cache);
+			nsvg__scanlineSolid(&r->bitmap[y * r->stride] + xmin*4, xmax-xmin+1, &r->scanline[xmin], xmin, y, tx,ty, sx,sy, cache);
 		}
 	}
 
@@ -1366,9 +1376,9 @@
 }
 */
 
-void nsvgRasterize(NSVGrasterizer* r,
-				   NSVGimage* image, float tx, float ty, float scale,
-				   unsigned char* dst, int w, int h, int stride)
+void nsvgRasterizeXY(NSVGrasterizer* r,
+				     NSVGimage* image, float tx, float ty, float sx, float sy,
+				     unsigned char* dst, int w, int h, int stride)
 {
 	NSVGshape *shape = NULL;
 	NSVGedge *e = NULL;
@@ -1398,7 +1408,7 @@
 			r->freelist = NULL;
 			r->nedges = 0;
 
-			nsvg__flattenShape(r, shape, scale);
+			nsvg__flattenShape(r, shape, sx,sy);
 
 			// Scale and translate edges
 			for (i = 0; i < r->nedges; i++) {
@@ -1416,14 +1426,14 @@
 			// now, traverse the scanlines and find the intersections on each scanline, use non-zero rule
 			nsvg__initPaint(&cache, &shape->fill, shape->opacity);
 
-			nsvg__rasterizeSortedEdges(r, tx,ty,scale, &cache, shape->fillRule);
+			nsvg__rasterizeSortedEdges(r, tx,ty, sx,sy, &cache, shape->fillRule);
 		}
-		if (shape->stroke.type != NSVG_PAINT_NONE && (shape->strokeWidth * scale) > 0.01f) {
+		if (shape->stroke.type != NSVG_PAINT_NONE && (shape->strokeWidth * sx) > 0.01f) {
 			nsvg__resetPool(r);
 			r->freelist = NULL;
 			r->nedges = 0;
 
-			nsvg__flattenShapeStroke(r, shape, scale);
+			nsvg__flattenShapeStroke(r, shape, sx,sy);
 
 //			dumpEdges(r, "edge.svg");
 
@@ -1443,7 +1453,7 @@
 			// now, traverse the scanlines and find the intersections on each scanline, use non-zero rule
 			nsvg__initPaint(&cache, &shape->stroke, shape->opacity);
 
-			nsvg__rasterizeSortedEdges(r, tx,ty,scale, &cache, NSVG_FILLRULE_NONZERO);
+			nsvg__rasterizeSortedEdges(r, tx,ty,sx,sy, &cache, NSVG_FILLRULE_NONZERO);
 		}
 	}
 
@@ -1455,6 +1465,13 @@
 	r->stride = 0;
 }
 
+void nsvgRasterize(NSVGrasterizer* r,
+				   NSVGimage* image, float tx, float ty, float scale,
+				   unsigned char* dst, int w, int h, int stride)
+{
+	nsvgRasterizeXY(r,image, tx, ty, scale, scale, dst, w, h, stride);
+}
+
 #endif // NANOSVGRAST_IMPLEMENTATION
 
 #endif // NANOSVGRAST_H

# source
https://github.com/memononen/nanosvg/archive/ea6a6aca009422bba0dbad4c80df6e6ba0c82183/nanosvg-20241219.tar.gz
